api/v1_users_verify_token.go

```diff
@@ -0,0 +1,162 @@
+
+	if len(signatureBytes) != 65 {
+		return fiber.NewError(fiber.StatusBadRequest, "The JWT signature could not be decoded")
+	}
+	if signatureBytes[64] >= 27 {
+		signatureBytes[64] -= 27
+	}
+
+	publicKey, err := crypto.SigToPub(finalHash.Bytes(), signatureBytes)
+	if err != nil {
+		return fiber.NewError(fiber.StatusNotFound, "The JWT signature is invalid - wallet could not be recovered")
+	}
+
+	recoveredAddr := crypto.PubkeyToAddress(*publicKey)
+	walletLower := strings.ToLower(recoveredAddr.Hex())
+
+	// 4. Check that user from payload matches the user from the wallet from the signature
+	paddedPayload := base64Payload
+	if len(paddedPayload)%4 != 0 {
+		paddedPayload += strings.Repeat("=", 4-len(paddedPayload)%4)
+	}
+
+	stringifiedPayload, err := base64.URLEncoding.DecodeString(paddedPayload)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "JWT payload could not be decoded")
+	}
+
+	var payload map[string]interface{}
+	if err := json.Unmarshal(stringifiedPayload, &payload); err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "JWT payload could not be decoded")
+	}
+
+	// Convert numeric fields to strings to match legacy behavior
+	if iat, exists := payload["iat"]; exists {
+		if iatNum, ok := iat.(float64); ok {
+			payload["iat"] = fmt.Sprintf("%.0f", iatNum)
+		}
+	}
+
+	// Get user ID from JWT payload
+	userIdInterface, exists := payload["userId"]
+	if !exists {
+		return fiber.NewError(fiber.StatusBadRequest, "JWT payload missing userId field")
+	}
+
+	userIdStr, ok := userIdInterface.(string)
+	if !ok {
+		return fiber.NewError(fiber.StatusBadRequest, "JWT payload userId must be a string")
+	}
+
+	jwtUserId, err := trashid.DecodeHashId(userIdStr)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "Invalid JWT payload userId")
+	}
+
+	// Get user ID associated with recovered wallet
+	walletUserId, err := app.queries.GetUserForWallet(c.Context(), walletLower)
+	if err != nil {
+		if err == pgx.ErrNoRows {
+			return fiber.NewError(fiber.StatusNotFound, "The JWT signature is invalid - invalid wallet")
+		}
+		return err
+	}
+
+	// Check if wallet user matches JWT user or if wallet user is a manager
+	if int32(walletUserId) != int32(jwtUserId) {
+		// Check if the signing user is a manager of the JWT user
+		isManager, err := app.isActiveManager(c.Context(), int32(jwtUserId), int32(walletUserId))
+		if err != nil {
+			return err
+		}
+
+		if !isManager {
+			return fiber.NewError(fiber.StatusForbidden, "The JWT signature is invalid - the wallet does not match the user")
+		}
+	}
+
+	// 5. Send back the decoded payload
+	return c.JSON(fiber.Map{
+		"data": payload,
+	})
+}
+
+func (app *ApiServer) isActiveManager(ctx context.Context, userId int32, managerUserId int32) (bool, error) {
+	grants, err := app.queries.GetGrantsForUserId(ctx, dbv1.GetGrantsForUserIdParams{
+		UserID:     userId,
+		IsRevoked:  false,
+		IsApproved: pgtype.Bool{Bool: true, Valid: true},
```

Comment:
oof this is awkward.

Reply:
yeah.... i didn't want to change your underlying sql model, probably the real fix is moving away from sqlc